src/backend/distributed/operations/worker_split_shard_replication_setup_udf.c

```diff
@@ -0,0 +1,448 @@
+ * children along with their placements. This info is required during the catch up
+ * phase of logical replication.
+ * This meta information is stored in a shared memory segment and accessed
+ * by logical decoding plugin.
+ *
+ * Split information is given by user as an Array of custom data type 'citus.split_shard_info'.
+ * (worker_split_shard_replication_setup(citus.split_shard_info[]))
+ *
+ * Fields of custom data type 'citus.split_shard_info':
+ * source_shard_id - id of the shard that is undergoing a split
+ *
+ * child_shard_id  - id of shard that stores a specific range of values
+ *                   belonging to sourceShardId(parent)
+ *
+ * shard_min_value - lower bound(inclusive) of hash value which childShard stores.
+ *
+ * shard_max_value - upper bound(inclusive) of hash value which childShard stores
+ *
+ * node_id         - Node where the childShardId is located
+ *
+ * The function parses the data and builds routing map with key for each distinct
+ * <nodeId, tableOwner> pair. Multiple shards can be placed on the same destination node.
+ * Source and destination nodes can be same too.
+ *
+ * There is a 1-1 mapping between a table owner and a replication slot. One replication
+ * slot takes care of replicating changes for all shards belonging to the same owner on a particular node.
+ *
+ * During the replication phase, WAL senders will attach to the shared memory
+ * populated by current UDF. It routes the tuple from the source shard to the appropriate destination
+ * shard for which the respective slot is responsible.
+ */
+Datum
+worker_split_shard_replication_setup(PG_FUNCTION_ARGS)
+{
+	if (PG_ARGISNULL(0))
+	{
+		ereport(ERROR, (errmsg("split_shard_info array cannot be NULL")));
+	}
+
+	ArrayType *shardInfoArrayObject = PG_GETARG_ARRAYTYPE_P(0);
+	if (array_contains_nulls(shardInfoArrayObject))
+	{
+		ereport(ERROR, (errmsg("Unexpectedly shard info array contains a null value")));
+	}
+
+	/* SetupMap */
+	ShardInfoHashMap = SetupHashMapForShardInfo();
+
+	int shardSplitInfoCount = 0;
+
+	ArrayIterator shardInfo_iterator = array_create_iterator(shardInfoArrayObject, 0,
+															 NULL);
+	Datum shardInfoDatum = 0;
+	bool isnull = false;
+	while (array_iterate(shardInfo_iterator, &shardInfoDatum, &isnull))
+	{
+		uint64 sourceShardId = 0;
+		uint64 childShardId = 0;
+		int32 minValue = 0;
+		int32 maxValue = 0;
+		int32 nodeId = 0;
+
+		ParseShardSplitInfoFromDatum(shardInfoDatum, &sourceShardId, &childShardId,
+									 &minValue, &maxValue, &nodeId);
+
+		ShardSplitInfo *shardSplitInfo = CreateShardSplitInfo(
+			sourceShardId,
+			childShardId,
+			minValue,
+			maxValue,
+			nodeId);
+
+		AddShardSplitInfoEntryForNodeInMap(shardSplitInfo);
+		shardSplitInfoCount++;
+	}
+
+	dsm_handle dsmHandle;
+	ShardSplitInfoSMHeader *splitShardInfoSMHeader =
+		CreateSharedMemoryForShardSplitInfo(shardSplitInfoCount, &dsmHandle);
+
+	PopulateShardSplitInfoInSM(splitShardInfoSMHeader,
+							   ShardInfoHashMap);
```

Comment:
'ShardInfoHashMap' is static so you don't need this param. Needs to be removed.

Reply:
Done. Thanks.